internal/runbits/projectmigration/projectmigration.go

```diff
@@ -22,31 +23,41 @@ type promptable interface {
 	Confirm(title, message string, defaultChoice *bool) (bool, error)
 }
 
-// PromptToMigrateIfNecessary checks if the given project has been migrated to the new project
-// format. If not, the user is prompted to do so. This should be called in main() as soon as a
-// project file is parsed.
-func PromptToMigrateIfNecessary(proj projecter, prompt promptable, out output.Outputer) error {
-	projectDir := filepath.Dir(proj.Path())
-	if _, err := localcommit.Get(projectDir); err == nil || !localcommit.IsFileDoesNotExistError(err) {
-		return err
+var prompt promptable
+var out output.Outputer
+var declined bool
+
+func Register(prompt_ promptable, out_ output.Outputer) {
+	prompt = prompt_
+	out = out_
+}
```

Comment:
This is an antipattern for how we do this type of thing in our codebase. Either we should pass around a pointed struct that contains this context, or we always require these variables to be passed when invoking PromptAndMigrate.

That said I can see why you went this route as it feels really wasteful to add all this boilerplate for effectively a legacy use-case.

I feel like we need a pattern for handling legacy transitions like this, but that's probably too much to handle within the scope of this story.

Perhaps for this story we should clearly call out its legacy nature. I'm thinking add comments here clearly explaining that this is an antipattern and it's only ok because this is for a legacy use-case. I'm also thinking we should clearly mark packages meant for legacy use-cases as such, maybe by putting them a dir deeper under a parent dir called "legacy"? So you'd have `runbits/legacy/projectmigration`.

I'm not a big fan of any of these solutions, but I don't see a solution that doesn't have downsides. Main things I want to achieve here is maintainable code that doesn't start infecting other code, which I think is also where you were coming from. Happy to iterate with other ideas.

Reply:
The need for this arose because we were auto-migrating the project as soon as it was read, which did not allow for any prompt/output parameter passing (i.e. `projectfile.FromPath(pjPath)` followed by `project.New(pjf, out)`), especially since analytics was not available yet (prompt needs this, and analytics needs to know the project; circular dependency).

However, after a couple of iterations, we're making prompting a lazy action, so we can pass around prompter and output interfaces as normal. It's quite verbose, but I've made the change.

From what I understand, there's no need for a `legacy/` parent package.